DC/OS Metrics Service

malnick opened the pull request:
## Description
### DC/OS Metrics Service
The PR enables the DC/OS metrics service. The service runs on a domain socket on both masters and agents. On masters it exposes host-level metrics and on agents it exposes host, container and metrics gathered from the Mesos Metrics module. 

### Mesos Metrics Module
This PR adds the Mesos metrics module to OSS DC/OS. Along with that addition are metrics module-specific CPP libraries for avro and boost. 

### Integration Tests
The integration test ensures that the metrics service runs and returns host level metrics for each DC/OS role. It ensures that metrics sent to STATSD_UDP_HOST/PORT inside a Mesos deployed container are collected by the metrics service and exposed with the expected values. 

## Code Coverage
### Summary
```
Packages   100% Files   100% Classes   100% Methods   86% Lines   63% Conditionals   100%
```
### Detailed Coverage
```
github.com/dcos/dcos-metrics/collectors/framework/framework.go:63:	New				100.0%
github.com/dcos/dcos-metrics/collectors/framework/framework.go:74:	Transform			78.9%
github.com/dcos/dcos-metrics/collectors/framework/framework.go:131:	RunFrameworkTCPListener		75.0%
github.com/dcos/dcos-metrics/collectors/framework/framework.go:156:	handleConnection		73.5%
github.com/dcos/dcos-metrics/collectors/framework/framework.go:228:	GetTopic			100.0%
github.com/dcos/dcos-metrics/collectors/framework/framework.go:244:	Read				80.0%
github.com/dcos/dcos-metrics/collectors/framework/record.go:41:		extract				83.9%
github.com/dcos/dcos-metrics/collectors/framework/record.go:108:	createObjectFromRecord		80.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:152:	RunPoller			0.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:162:	pollMesosAgent			0.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:182:	getContainerMetrics		100.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:198:	getAgentState			100.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:212:	transformContainerMetrics	81.8%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:248:	getFrameworkInfoByFrameworkID	100.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:260:	getLabelsByContainerID		100.0%
github.com/dcos/dcos-metrics/collectors/mesos/agent/agent.go:278:	buildDatapoints			65.6%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:79:		getNodeMetrics			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:125:		init				100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:132:		getUptime			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:137:		getProcessCount			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:142:		getNumCores			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:151:		getLoadAvg			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:156:		getCPUTimes			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:167:		getMemory			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:172:		getSwap				100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:177:		getFilesystems			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:195:		getNetworkInterfaces		100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:215:		calculatePcts			100.0%
github.com/dcos/dcos-metrics/collectors/node/metrics.go:236:		round				100.0%
github.com/dcos/dcos-metrics/collectors/node/node.go:46:		RunPoller			0.0%
github.com/dcos/dcos-metrics/collectors/node/node.go:57:		pollHost			0.0%
github.com/dcos/dcos-metrics/collectors/node/node.go:76:		transform			100.0%
github.com/dcos/dcos-metrics/collectors/node/node.go:99:		buildDatapoints			71.9%
github.com/dcos/dcos-metrics/config.go:85:				setFlags			100.0%
github.com/dcos/dcos-metrics/config.go:92:				loadConfig			66.7%
github.com/dcos/dcos-metrics/config.go:105:				getNodeInfo			0.0%
github.com/dcos/dcos-metrics/config.go:147:				newConfig			100.0%
github.com/dcos/dcos-metrics/config.go:171:				getNewConfig			0.0%
github.com/dcos/dcos-metrics/dcos-metrics.go:35:			main				0.0%
github.com/dcos/dcos-metrics/dcos-metrics.go:124:			producerIsConfigured		100.0%
github.com/dcos/dcos-metrics/producers/http/handlers.go:27:		nodeHandler			64.3%
github.com/dcos/dcos-metrics/producers/http/handlers.go:50:		containersHandler		57.1%
github.com/dcos/dcos-metrics/producers/http/handlers.go:73:		containerHandler		66.7%
github.com/dcos/dcos-metrics/producers/http/handlers.go:91:		containerAppHandler		70.0%
github.com/dcos/dcos-metrics/producers/http/handlers.go:110:		containerAppMetricHandler	61.9%
github.com/dcos/dcos-metrics/producers/http/handlers.go:147:		pingHandler			100.0%
github.com/dcos/dcos-metrics/producers/http/handlers.go:160:		encode				50.0%
github.com/dcos/dcos-metrics/producers/http/http.go:51:			New				100.0%
github.com/dcos/dcos-metrics/producers/http/http.go:63:			Run				95.7%
github.com/dcos/dcos-metrics/producers/http/http.go:123:		janitor				100.0%
github.com/dcos/dcos-metrics/producers/http/logger.go:22:		logger				100.0%
github.com/dcos/dcos-metrics/producers/http/router.go:23:		loadRoutes			100.0%
github.com/dcos/dcos-metrics/producers/http/router.go:41:		newRouter			100.0%
github.com/dcos/dcos-metrics/util/http/client/client.go:38:		Fetch				50.0%
github.com/dcos/dcos-metrics/util/http/helpers/helpers.go:30:		loadCAPool			81.8%
github.com/dcos/dcos-metrics/util/http/helpers/helpers.go:51:		getTransport			88.9%
github.com/dcos/dcos-metrics/util/http/helpers/helpers.go:75:		NewMetricsClient		0.0%
total:									(statements)			63.1%
```

malnick commented:
@cmaloney This needs a WIP label 

CC @nickbp @dcos/day2ops

adam-mesos commented:
Let's make sure this doesn't get into 1.8.5, and only in 1.9/1.10

nickbp commented on packages/avro-cpp/buildinfo.json:
```diff
@@ -0,0 +1,8 @@
+  "requires": ["mesos", "boost-libs"],
+  "single_source": {
+    "kind": "url_extract",
+    "url": "http://apache.mirrors.pair.com/avro/avro-1.8.0/cpp/avro-cpp-1.8.0.tar.gz",
```
Could use this new URL mirror for this file: https://downloads.mesosphere.com/pkgpanda-artifact-cache/avro-cpp-1.8.0.tar.gz

malnick replied:
I updated the line but this comment still shows the previous commits :/ Did github comments just become less awesome?

cmaloney commented:
@adam-mesos things have to be tagged milestone 1.8.5 to be backported. By default things are going to 1.9

nickbp commented on packages/dcos-metrics-collector/build:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
The collector needs sufficient credentials to poll the local mesos agent. When being run manually they're configured via either `--auth_credential` or `AUTH_CREDENTIAL` envvar. Is it feasible to add that here? (preconfigured service account?)

nickbp replied:
Alternately if credentials arent needed to poll mesos agents then this is a moot point.

malnick replied:
Only required on EE

malnick replied:
Will have to add that in the EE PR eventually

mellenburg commented on packages/dcos-metrics-collector/build:
```diff
@@ -0,0 +1,18 @@
+export GOPATH=$GOPATH:/pkg
+mkdir -p /pkg/src/github.com/dcos/dcos-metrics
+# Create the GOPATH for the go tool to work properly
+cp -R /pkg/src/* /pkg/src/github.com/dcos/dcos-metrics/
```
Heres how you can do this:
https://github.com/dcos/dcos/blob/master/packages/3dt/build#L5

malnick replied:
https://github.com/dcos/dcos/pull/778/commits/810f863d2cd7c2513c965fbf531db4e1d87659b6

That didn't seem to work either.

mellenburg replied:
Look again:
You have dcos-metrics but need dcos-metrics-collector

malnick replied:
The repo is called dcos-metrics though?

malnick replied:
We're building a sub-project of this repo

alberts commented on packages/adminrouter/buildinfo.json:
```diff
@@ -8,9 +8,9 @@
-          "git": "https://github.com/dcos/adminrouter.git",
-          "ref": "0f1e4d44701d5b9960da980f8136c03c98c96594",
-          "ref_origin": "master"
+          "git": "https://github.com/malnick/adminrouter.git",
```
this needs to go upstream eventually

malnick replied:
Referenced in https://github.com/dcos/adminrouter/pull/20

nickbp commented on packages/mesos-metrics-module/buildinfo.json:
```diff
@@ -0,0 +1,9 @@
+  "requires": ["mesos", "boost-libs", "avro-cpp"],
+  "single_source" : {
+    "kind": "git",
+    "git": "git@github.com:mesosphere/dcos-stats.git",
```
Wrong repo?

malnick replied:
We haven't touched this yet, still redirects but definitly needs to change soon.

rji commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,130 @@
+    def expected_datapoint_response(response):
+        for datapoint in response["datapoints"]:
+            if datapoint["name"] not in expected_node_metric_datapoints:
+                return False
```
I think the logic is backwards here. This test will return `False` if it finds any datapoints _outside_ the `expected_node_metric_datapoints`, instead of ensuring that the minimum required datapoints exist. Consider the following:

We add a new feature and `response["datapoints"]` contains a datapoint with the `name` "foo". "foo" is not in `expected_node_metric_datapoints`, so this test returns `False` (as expected). But we also want to ensure that a datapoint with the `name` `load.15min` exists. There isn't any logic here to ensure `response["datapoints"]` contains all of the elements of `expected_node_metric_datapoints`.

A better implementation might be:

```python
for datapoint in expected_node_metric_datapoints:
    if datapoint not in response["datapoints"]:
        return False
return True
```

This ensures we're returning the minimum required datapoints, while not breaking this test if we add additional datapoints in the future. Of course, if you want to ensure the minimum required datapoints are there _and_ we haven't added anything new, you could simply check the length of `response["datapoints"]` against the length of `expected_node_metric_datapoints`.

malnick replied:
We do a similar test for signal, where if you add a new package to DC/OS, we want to make sure signal is picking it up, so we test every single service exists. I'd like to do the same thing here, that way we can be sure the metrics we get in the response are the ones we intend to set. 

If we only check the length, the length could be anything, and if we add a new service, we'll have to update the test anyways. If we're going to update the test anyways, we come full circle and we're back to just adding the name of the thing we're testing.

rji replied:
> that way we can be sure the metrics we get in the response are the ones we intend to set.
That was my point :smile:

> If we only check the length, the length could be anything
I didn't mean to suggest to only check the length, but to check the length in addition to the mandatory / required datapoints in the response. You could check the length to ensure that no _new_ datapoints have been added that you weren't expecting to see in this test.

malnick replied:
+1 I'll add it, thanks.

mellenburg commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,129 @@
+        "mem": 128.0,
+        "instances": 1
+    }
+    app_endpoints = cluster.marathon.deploy_app(marathon_json)
```
use 
```
with cluster.marathon_deploy_and_cleanup:
    # same stuff
```

malnick replied:
Why cleanup?

malnick commented:
@spahl This is ready for review.

rji commented on gen/dcos-config.yaml:
```diff
@@ -379,6 +379,31 @@ package:
+        node:
+          poll_period: 15
+
+        # DC/OS (Mesos) agent configuration
```
Do we need master configuration as well?

malnick replied:
It doesn't query the master yet.

cmaloney commented:
High level comments
 - PR Title needs to be updated if this is no longer WIP
 - The PR top level needs to be updated to use https://github.com/dcos/dcos/blob/master/PULL_REQUEST_TEMPLATE.md

malnick commented:
@cmaloney The description and title are updated.

cmaloney commented on packages/boost-libs/build:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Mesos already includes a whole bunch of boost, what are we using which isn't bundled by mesos? Can we just use mesos bundling?

If we are going to do this outside of mesos, need to build it ones and configure mesos to use it.

malnick replied:
@nickbp Can you clarify here?

@cmaloney This was copied out of our enterprise distro to OSS the project, I'm not sure what the right answer is on this.

nickbp replied:
- The metrics module only directly pulls in `system` as `asio` requires it (despite otherwise being header-only).
- The avro library pulls in the other three (`filesystem`/`iostreams`/`program_options`).
The mesos module meanwhile is just using header-only components IIRC

alberts replied:
As far as I can tell, this was answered.

lingmann replied:
Looks like this was addressed in a separate PR bundling boost-libs: #1030

cmaloney commented on test_util/helpers.py:
```diff
@@ -57,7 +57,7 @@ def __init__(self, default_host_url, api_base, default_headers=None, ca_cert_pat
         self.default_headers = default_headers
         self.ca_cert_path = ca_cert_path
 
-    def api_request(self, method, path, host_url=None, **kwargs):
```
this change / the wrapper change should not be in this PR

cmaloney commented on packages/mesos-metrics-module/extra/metrics.json:
```diff
@@ -0,0 +1,23 @@
+            {"key": "output_collector_chunk_size_datapoints", "value": "100"},
+            {"key": "output_collector_chunk_timeout_seconds", "value": "10"},
+            {"key": "state_path_dir",
+             "value": "/var/run/mesos/isolators/com_mesosphere_MetricsIsolatorModule/"}
```
Why is this doing `/var/run/mesos`, all runtime state should be in subdirectories of `/var/run/dcos`.

cmaloney commented on packages/dcos-metrics/extra/dcos-metrics-master.service:
```diff
@@ -0,0 +1,10 @@
+User=dcos_metrics
+EnvironmentFile=/opt/mesosphere/environment
+ExecStartPre=/opt/mesosphere/bin/bootstrap dcos-metrics
+ExecStart=/opt/mesosphere/bin/dcos-metrics -config /opt/mesosphere/etc/dcos-metrics-config.yaml -role master -loglevel debug 
```
why is debug log level appropriate for production?

malnick replied:
It's not, will fix.

cmaloney commented on gen/dcos-config.yaml:
```diff
@@ -379,6 +379,31 @@ package:
+          poll_period: 15
+
+        # DC/OS (Mesos) agent configuration
+        mesos_agent:
```
This same config file is used both for masters and agents. Does metrics magically select whether to ignore this section if it's an agent? Or should we be outputting two config files, one for agents which says to ask the agent, one for masters which says not to ask it currently.

cmaloney replied:
For this also: What would this all look like if I were running a master and agent on the same host? (Something which is planned for the future)

malnick replied:
In master mode it only exposes host-level metrics, so it does not make requests to mesos-agent API, and therefore ignores this configuration if present.

cmaloney commented on packages/mesos-metrics-module/buildinfo.json:
```diff
@@ -0,0 +1,9 @@
+  "requires": ["mesos", "boost-libs", "avro-cpp"],
+  "single_source" : {
+    "kind": "git",
+    "git": "git@github.com:dcos/dcos-metrics.git",
```
The dcos-metrics package already clones and builds this repo, having two different pkgpanda packages cloning the same repository just makes them really likely to get out of sync. Building the mesos module and building the other metrics components all of which are built from the same commit of this same repository should happen inside one pkgpanda package.

cmaloney commented on packages/avro-cpp/build:
```diff
@@ -0,0 +1,25 @@
+  ..
+
+# build/unittest
+make -j8
```
This needs to use `make -j$NUM_CORES` (see: https://github.com/dcos/dcos/commit/c55b0c82c3c74d1e5c37b7d944d527bcda6e5dd9)

cmaloney commented on packages/mesos-metrics-module/build:
```diff
@@ -0,0 +1,31 @@
+  -DUSE_LOCAL_AVRO=False \
+  /pkg/src/mesos-metrics-module/mesos_module/
+
+make -j8
```
This needs to use `make -j$NUM_CORES` (see: https://github.com/dcos/dcos/commit/c55b0c82c3c74d1e5c37b7d944d527bcda6e5dd9)

cmaloney commented on packages/mesos-metrics-module/build:
```diff
@@ -0,0 +1,31 @@
+# install lib to $PKG_PATH/...
+make install
+
+popd
```
there isn't any value in doing the popd here (all we do is copy one file after this with absolute paths everywhere....)

cmaloney commented on packages/mesos-metrics-module/build:
```diff
@@ -0,0 +1,31 @@
+  -Ddefault_BIN_DIR=/opt/mesosphere/bin \
+  -Ddefault_LIB_DIR=/opt/mesosphere/lib \
+  -DUSE_LOCAL_PROTOBUF=False \
+  -Dprotobuf_INCLUDE_DIR=/opt/mesosphere/lib/mesos/3rdparty/include \
```
Since this is a mesos module, why can't we just use a configuration flag which says "here's all the mesos bundled includes" rather than having to call them out one by one?

malnick replied:
@nickbp CC

nickbp replied:
There's currently `HINTS` entries which default to `${mesos_INCLUDE_DIR}`. Could add a second HINT which points to `${default_LIB_DIR}/mesos/3rdparty/include`:
- glog: https://github.com/dcos/dcos-metrics/blob/master/mesos_module/CMakeLists.txt#L148
- protobuf: https://github.com/dcos/dcos-metrics/blob/master/mesos_module/CMakeLists.txt#L124

see also: https://cmake.org/cmake/help/v3.0/command/find_path.html

cmaloney commented on packages/dcos-metrics/extra/dcos-metrics.socket:
```diff
@@ -0,0 +1,4 @@
+[Unit]
+Description=Metrics Socket: Socket for DC/OS Metrics Service
+[Socket]
+ListenStream=/run/dcos/dcos-metrics.sock
```
According to systemd `For each socket file, a matching service file must exist, describing the service to start on incoming traffic on the socket`. which means this needs to have the name `dcos-metrics-master.socket` on masters, `dcos-metrics-agent.socket` on agents I think

cmaloney commented on packages/dcos-metrics/extra/dcos-metrics-master.service:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Metrics Service: DC/OS Metrics Service for Application, Container and Host Level Metrics.
```
This description should be different between master / agent so people can tell which they're looking at in the ui

cmaloney commented on packages/dcos-metrics/build:
```diff
@@ -0,0 +1,40 @@
+mv /pkg/src/github.com/dcos/dcos-metrics/build/statsd-emitter/dcos-metrics-statsd-emitter-* $PKG_PATH/bin/statsd-emitter
+
+# Create the service file for all roles 
+agent_service="$PKG_PATH/dcos.target.wants_slave/dcos-metrics.service"
```
Two systemd services with different contents should never have the same name when we're shipping them. renaming them all to dcos-metrics.service will make debugging difficult on hosts (people assume same name -> same contents).

cmaloney commented on packages/adminrouter/buildinfo.json:
```diff
@@ -9,7 +9,7 @@
           "kind": "git",
           "git": "https://github.com/dcos/adminrouter.git",
-          "ref": "0f1e4d44701d5b9960da980f8136c03c98c96594",
+          "ref": "a6f1388f77c6e0324e1d2ce1c7490d8ddf90970d",
```
nit: This pulls in a lot of unrelated changes to adminrouter, almost none of which have tests.

malnick commented:
Rebased on #998

malnick commented:
@cmaloney Comments addressed, ready for another pass. TIA.

alberts commented on gen/dcos-config.yaml:
```diff
@@ -379,6 +379,31 @@ package:
       }
+  - path: /etc/dcos-metrics-config.yaml
+    content: |
+      #######################################
```
nit: drop this comment... and probably the others. I don't think they add any value.

malnick replied:
Fixed.

alberts commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,128 @@
+    """
+    for agent in cluster.slaves:
+        response = cluster.metrics.get('ping', node=agent)
+        assert response.status_code == 200, 'Got non-200 status code '
```
we're allowed up to 120 these days...

malnick replied:
Yeah but ping should definitely return only a 200.

alberts replied:
i mean, we can make code be 120 columns wide

malnick replied:
Fixed.

alberts commented on test_util/cluster_api.py:
```diff
@@ -331,6 +331,10 @@ def marathon(self):
 
+    @property
+    def metrics(self):
+        return self.get_client('/system/v1/metrics/v0')
```
that's a lot of versions...

malnick replied:
There really isn't an emoticon that describes my emotion for this statement.

malnick commented:
Updated description to reflect latest unit test coverage by @rji and rebased commit history to something that looks reasonable. This is ready for another pass. @cmaloney @alberts @spahl

alberts commented on gen/calc.py:
```diff
@@ -557,7 +557,8 @@ def calculate_set(parameter):
         'minuteman_min_named_ip_erltuple': calculate_minuteman_min_named_ip_erltuple,
         'minuteman_max_named_ip_erltuple': calculate_minuteman_max_named_ip_erltuple,
-        'mesos_isolation': calculate_mesos_isolation,
+        'mesos_isolation': 'cgroups/cpu,cgroups/mem,disk/du,network/cni,filesystem/linux,docker/runtime,docker/volume'
```
why'd we nuke this?

malnick replied:
Looks like something moved in underneath me in a rebase somewhere. @nickbp @cmaloney Looks like the method returns one set of isolators and this is appending the metrics isolator. Do you think I can simply append the metrics isolator to what is being returned in the method, similarly to what we do here?

malnick replied:
I updated the isolation scheme from the method to append the metrics isolation in the returned list. This was modified when pods was merged last week, which is where the rebase broke down for me earlier this week.

alberts commented on packages/dcos-metrics/buildinfo.json:
```diff
@@ -0,0 +1,11 @@
+    "git": "https://github.com/dcos/dcos-metrics.git",
+    "ref": "b1a88554bf107f9443245e787e82db4f502b300d",
+    "ref_origin": "master"
+
```
blank line

malnick replied:
Fixed

rji commented on packages/dcos-metrics/buildinfo.json:
```diff
@@ -0,0 +1,10 @@
+  "single_source" : {
+    "kind": "git",
+    "git": "https://github.com/dcos/dcos-metrics.git",
+    "ref": "b1a88554bf107f9443245e787e82db4f502b300d",
```
Can we bump this to `3b07a7cac98153e7f8708b84538953f3f83f9447`

rji replied:
Or if `ref` will accept a tag, the repo is currently tagged at `0.0.8`, which will hopefully be the last tag before 1.0-rc1.

cmaloney commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,128 @@
+LATENCY = 60
+
+
+@retrying.retry(wait_fixed=2000, stop_max_delay=LATENCY * 1000)
```
why do we have to retry this? What is failing that makes it so we can't know deterministically what the results are?

This retry loop is also retrying if it fails for _any_ reason (including that the asserts in the test failed). If something needs to be retried we have to be far more precise. The retry time in a running cluster should also be far less than 60 seconds.

lingmann replied:
@malnick can you address this question? I see we still have some retrying logic in the PR / not sure if this was ever addressed or discussed.

cmaloney commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,128 @@
+    for agent in cluster.slaves:
+        response = cluster.metrics.get('node', node=agent)
+
+        assert response.status_code == 200, 'Got non-200 status code '
```
assert r.status_code == 200 ends up producing really hard to debug / track down errors when it fails (you don't get the status code or response body printed which are both usually critical in solving). Please use r.raise_for_status() or do a `assert r.ok, "status_code: {}, content: {}".format(r.status_code, r.content))`. This goes for this instance and all the other ones. The "Got non-200 status code" is also not very informative when this fails.

cmaloney commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,128 @@
+                return False
+            return True
+
+    for agent in cluster.slaves:
```
there isn't a lot of value being added by throwing all these loops in one big `test_`. Why can't there be a `test_masters` / `test_agents`, etc.?

malnick commented on gen/dcos-config.yaml:
```diff
@@ -379,6 +379,23 @@ package:
       }
+  - path: /etc/dcos-metrics-config.yaml
+    content: |
+      collector:
```
That piece of configuration is set via the command line flag `--role <agent,master>`. 

I'm not sure what you mean by "just defaults". The collector is broken down into discrete collectors and producers. Today, we have a node, mesos agent, and application (container metrics sent from metrics module) collector. We have a single producer for HTTP, but that will increase in 1.10. So this config file reflects that design.

malnick commented:
@cmaloney This is ready for another pass. Appears that rebasing might break stuff (see Slack conversation) so I'm holding off on that until I see it's stable.

malnick commented:
TODO

- [ ] Merge https://github.com/dcos/dcos-metrics/pull/50 and point buildinfo for dcos-metrics to bumped release tag
- [ ] @spahl Is doing the boost lib refactoring in a separate PR (TBD)

cmaloney commented:
This needs to be rebased on master so the boost-libs pull out isn't included in this

spahl commented:
https://github.com/dcos/dcos/pull/1030 is not in master yet. Putting them in the same train is simpler.

cmaloney commented:
hand-auditing the commits are identical is hard and tedious

spahl commented:
I'll manage. And trains get their own set of reviews as well.  I'd rather not wait 48 more hours for this.

spahl commented:
@malnick next time rebase vs cherry picking an extracted pr please.  It helps keeping the shas

spahl commented on pkgpanda/docker/dcos-builder/Dockerfile:
```diff
@@ -15,6 +15,7 @@ RUN apt-get -qq update && apt-get -y install \
   gettext-base \
   git \
   gzip \
+  libc6-dev \
```
Was about to put this on a train and saw this. What is this for @malnick?

malnick replied:
Misorderd list from golang:1.7 - it's on master, in another order :/

spahl commented:
once https://github.com/dcos/dcos/pull/1054 lands lets rebase

malnick commented:
Removed a401a3db9372f53e1c5bf1c55c3f26436da531b4 - needs rebase once 108 merges.

lingmann commented on packages/dcos-metrics/extra/metrics.json:
```diff
@@ -0,0 +1,23 @@
+            {"key": "output_collector_chunk_size_datapoints", "value": "100"},
+            {"key": "output_collector_chunk_timeout_seconds", "value": "10"},
+            {"key": "state_path_dir",
+             "value": "/var/run/mesos/isolators/com_mesosphere_MetricsIsolatorModule/"}
```
I believe runtime state should now all be stored in /var/run/dcos/

spahl replied:
I will change this in a separate PR and link it here

lingmann commented on packages/dcos-integration-test/extra/test_metrics.py:
```diff
@@ -0,0 +1,103 @@
+        assert response.json()['ok'], 'Status code: {}, Content {}'.format(response.status_code, response.content)
+
+
+@retrying.retry(wait_fixed=2000, stop_max_delay=LATENCY * 1000)
```
From an earlier PR review: 

> why do we have to retry this? What is failing that makes it so we can't know deterministically what the results are?
> 
> This retry loop is also retrying if it fails for any reason (including that the asserts in the test failed). If something needs to be retried we have to be far more precise. The retry time in a running cluster should also be far less than 60 seconds.

spahl commented:
Adminrouter changes are documented here for a later sha: https://github.com/dcos/dcos/pull/1056

adam-mesos commented:
Merged in Train 109 #1057 , please close
